src_lockfree/skiplist.ml

```diff
@@ -0,0 +1,337 @@
+    nodes with marked references along the way, and stopping as soon as the node
+    is found. *)
+let[@inline] find_node t key =
+  let level = Array.length t.root - 1 in
+  let prev_at_level = Array.unsafe_get t.root level in
+  find_node t key t.root prev_at_level level (Atomic.get prev_at_level)
+
+(* *)
+
+let create ?n_way ?(max_height = 10) ?(compare = Stdlib.compare) () =
+  (* The upper limit of [30] comes from [Random.bits ()] as well as from
+     limitations with 32-bit implementations.  It should not be a problem in
+     practice. *)
+  if max_height < 1 || 30 < max_height then
+    invalid_arg "Skiplist: max_height must be in the range [1, 30]";
+  let root = Array.init max_height @@ fun _ -> Atomic.make (Link Null) in
+  let size = Size.create ?n_way () in
+  { compare; root; size }
+
+(* *)
+
+let find_opt t key =
+  match find_node t key with Null -> None | Node r -> Some r.value
+
+(* *)
+
+let mem t key = match find_node t key with Null -> false | Node _ -> true
+
+(* *)
+
+let rec add t key value preds succs =
+  (not (find_path t key preds succs 0))
+  &&
+  let next =
+    let height = get_random_height (Array.length t.root) in
+    Array.init height @@ fun i -> Atomic.make (Link (Array.unsafe_get succs i))
+  in
+  let incr = Size.new_once t.size ~incr:true in
+  let (Node r as node : (_, _, [ `Node ]) node) =
+    Node { key; value; incr; next }
+  in
+  if
+    let succ = Link (Array.unsafe_get succs 0) in
+    Atomic.compare_and_set (Array.unsafe_get preds 0) succ (Link node)
+  then begin
+    if r.incr != Size.used_once then begin
+      Size.update_once t.size r.incr;
+      r.incr <- Size.used_once
+    end;
+    (* The node is now considered as added to the skiplist. *)
+    let rec update_levels level =
+      if Array.length next = level then begin
+        if is_marked (Atomic.get (Array.unsafe_get next (level - 1))) then begin
+          (* The node we finished adding has been removed concurrently.  To
+             ensure that no references we added to the node remain, we call
+             [find_node] which will remove nodes with marked references along
+             the way. *)
+          find_node t key |> ignore
+        end;
+        true
+      end
+      else if
+        let succ = Link (Array.unsafe_get succs level) in
+        Atomic.compare_and_set (Array.unsafe_get preds level) succ (Link node)
+      then update_levels (level + 1)
+      else
+        let _found = find_path t key preds succs level in
+        let rec update_nexts level' =
+          if level' < level then update_levels level
+          else
+            let next = Array.unsafe_get next level' in
+            match Atomic.get next with
+            | Link (Null | Node _) as before ->
+                let succ = Link (Array.unsafe_get succs level') in
+                if before != succ then
+                  if Atomic.compare_and_set next before succ then
+                    update_nexts (level' - 1)
```

Comment:
I don't see a reason to not do an `Atomic.set` here instead of and `Atomic.compare_and_set` as no one can access these levels of the added node yet.

Reply:
Hmm... It is a subtle thing, but after the node has been added to level 0, i.e. the first `compare_and_set` in `add` has succeeded, the node is consider as having been added to the skiplist and then `remove` may concurrently find and mark all the outgoing links from the node.  The `remove` operation is special in this sense. 
 Other operations cannot access outgoing links of a node on levels that the node has not yet been linked to other nodes of the skiplist.

Reply:
Oh, right, the node can be marked. The tests passed on my computer with this change so there is definitely a `dscheck` test missing here.